fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/cascades/values/RecordConstructorValueTest.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * RecordConstructorValueTest.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.values;
+
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.typing.TypeRepository;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.Message;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Test cases to test {@link RecordConstructorValue#deepCopyIfNeeded}.
+ */
+public class RecordConstructorValueTest {
+
+    private static final String[] suites = new String[] {"SPADES", "HEARTS", "DIAMONDS", "CLUBS"};
+
+    private static Type.Enum getCardsEnum() {
+        final var enumValues = new ArrayList<Type.Enum.EnumValue>();
+        for (var i = 0; i < suites.length; i++) {
+            enumValues.add(new Type.Enum.EnumValue(suites[i], i));
+        }
+        return new Type.Enum(false, enumValues, "enumType");
+    }
+
+    @Test
+    public void deepCopyMessageWithEnumTest() {
+        final var enumType = getCardsEnum();
+        var type = Type.Record.fromFieldsWithName("simpleType", false, List.of(Type.Record.Field.of(enumType, Optional.of("suit"))));
+        var repo = TypeRepository.newBuilder();
+        enumType.defineProtoType(repo);
```

Comment:
I think this line (56) can be removed because the call below will cause the type repository to recursively add the `enumType` since it is referenced by `simpleType` field in `type`.

same observation below at lines 83 and 100.

Reply:
removed